### PR TITLE
debug: fix edge cases in macOS debug symbol lookup

### PR DESCRIPTION
**Pull request by schmee**
This commit fixes two related things:

1. If the loop goes all the way through the slice without a match, on
   the last iteration `mid == symbols.len - 1` which causes
   `&symbols[mid + 1]` to be out of bounds. Handle this by clamping
   the value to the length of the slices.

2. If the address we're looking for is greater than the address of the
   last symbol in the `symbols` slice, we now match it to that symbol. Previously,
   we would miss this case since we only matched if the address was _in
   between_ the address of two symbols.

Closes https://github.com/ziglang/zig/issues/10666.

### CODE REVIEW

File: lib/std/debug.zig [hunk 1]
```diff
@@ -563,7 +563,7 @@ fn machoSearchSymbols(symbols: []const MachoSymbol, address: usize) ?*const Mach
     while (min < max) {
         const mid = min + (max - min) / 2;
         const curr = &symbols[mid];
-        const next = &symbols[mid + 1];
+        const next = &symbols[@minimum(mid + 1, symbols.len - 1)];
```
Comment (kubkon) [inline lib/std/debug.zig:566]: Out of curiosity, did you check if the same effect can be achieved by initially setting `max = len - 1` instead of `max = len`?

**Review by kubkon — CHANGES_REQUESTED** [review]
I think we ought to have an inline test of `machoSearchSymbols` to future proof any changes to it, etc. Is this something you'd be willing to help with, ideally as part of this PR?

File: lib/std/debug.zig (hunk 1 above)
Comment (schmee) [reply lib/std/debug.zig:566]: IIRC I did try that and it didn't work for some reason, let me try again to confirm 👍

**Comment by schmee** [comment]
@kubkon sure, I'll see what I can cook up!

File: lib/std/debug.zig (hunk 1 above)
Comment (schmee) [reply lib/std/debug.zig:566]: Seems like this does work, probably I tried it before I added the second half of the fix. Updated 👍

**Comment by schmee** [comment]
@kubkon updated and added tests.

File: lib/std/debug.zig [hunk 2]
```diff
@@ -581,9 +582,36 @@ fn machoSearchSymbols(symbols: []const MachoSymbol, address: usize) ?*const Mach
...
 }
 
+test "machoSearchSymbols" {
+    const symbols = [_]MachoSymbol{
+        .{ .addr = 100, .strx = undefined, .size = undefined, .ofile = undefined },
+        .{ .addr = 200, .strx = undefined, .size = undefined, .ofile = undefined },
+        .{ .addr = 300, .strx = undefined, .size = undefined, .ofile = undefined },
+    };
+
+    try testing.expectEqual(@as(?*const MachoSymbol, null), machoSearchSymbols(&symbols, 0));
```
Comment (kubkon) [inline lib/std/debug.zig:600]: Nit for the future: you could also just do

```zig
try testing.expect(machoSearchSymbols(&symbols, 99) == null);
```

**Review by kubkon — APPROVED** [review]
Looks great! Thank you so much for further investigating my suggestions @schmee!
